From 91f093333f09f4a402cbcd310976001df69a7012 Mon Sep 17 00:00:00 2001
From: Mike Snitzer <snitzer@redhat.com>
Date: Wed, 18 Jun 2014 20:04:20 -0400
Subject: [block] null_blk: fix softirq completions for queue_mode == 1

Message-id: <1403121864-7956-4-git-send-email-snitzer@redhat.com>
Patchwork-id: 84266
O-Subject: [RHEL7.1 PATCH 154/150] null_blk: fix softirq completions for queue_mode == 1
Bugzilla: 1105204
RH-Acked-by: David Milburn <dmilburn@redhat.com>
RH-Acked-by: Jeff Moyer <jmoyer@redhat.com>

BZ: 1105204

Upstream commit d891fa70876b37941a5c5bed813e73beb53ebcf7
Author: Jens Axboe <axboe@fb.com>
Date:   Mon Jun 16 11:40:25 2014 -0600

    null_blk: fix softirq completions for queue_mode == 1

    Only blk-mq completions have payload attached to the request, for
    request_fn mode we have stored it in req->special. This fixes an
    oops with queue_mode=1 and softirq completions.

    Signed-off-by: Jens Axboe <axboe@fb.com>

Signed-off-by: Jarod Wilson <jarod@redhat.com>

diff --git a/drivers/block/null_blk.c b/drivers/block/null_blk.c
index 98bb1f0..d58bf88 100644
--- a/drivers/block/null_blk.c
+++ b/drivers/block/null_blk.c
@@ -227,7 +227,10 @@ static void null_cmd_end_timer(struct nullb_cmd *cmd)
 
 static void null_softirq_done_fn(struct request *rq)
 {
- end_cmd(blk_mq_rq_to_pdu(rq));
+ if (queue_mode == NULL_Q_MQ)
+  end_cmd(blk_mq_rq_to_pdu(rq));
+ else
+  end_cmd(rq->special);
 }
 
 static inline void null_handle_cmd(struct nullb_cmd *cmd)
-- 
1.7.1